libimage/push.go

```diff
@@ -84,6 +86,12 @@ func (r *Runtime) Push(ctx context.Context, source, destination string, options
 		destRef = dockerRef
 	}
 
+	// docker-archive and only DockerV2Schema2MediaType support Gzip compression
+	if options.CompressionFormat != nil &&
+		(destRef.Transport().Name() == "docker-archive" || options.ManifestMIMEType == manifest.DockerV2Schema2MediaType) {
```

Comment:
This should also check for `docker-daemon` (and, uh, `ostree`, but transport is deprecated and not built by default, so let’s not).